### PR TITLE
[refactor] Program::this_thread_config() -> Program::compile_config()

### PR DESCRIPTION
**Pull request by PGZXB**
Issue: #7002 
Removed multi-thread version `Program::this_thread_config()` (see https://github.com/taichi-dev/taichi/pull/7159#issue-1532269774)

### CODE REVIEW

File: taichi/program/program.h [hunk 1]
```diff
@@ -121,22 +115,9 @@ class TI_DLL_EXPORT Program {
...
-      configs[thread_id] = configs[main_thread_id_];
-      return configs[thread_id];
-    }
-    return configs[thread_id];
-  }
-
-  const CompileConfig &config() {
-    return configs[main_thread_id_];
+  // TODO(PGZXB): Return `const CompileConfig &` (blocked by some tests)
+  CompileConfig &compile_config() {
```
Comment (ailzhang) [inline taichi/program/program.h:119]: nit: is it possible that we have another method that returns a const reference, and make sure only the problematic tests use this non-const version? Thanks!

Comment (PGZXB) [reply taichi/program/program.h:119]: Maybe we can provide `Program::set_compile_config`(only for tests)?
p.s. And I will refactor those tests to remove the temporary `Program::set_compile_config`.

Comment (ailzhang) [reply taichi/program/program.h:119]: sg!

Comment (PGZXB) [reply taichi/program/program.h]: I found that the dependencies on the non-const version in tests can be removed directly, so I killed them in this PR and fixed the ` CompileConfig &compile_config()` to ` const CompileConfig &compile_config()`. Please review again.

**Review by ailzhang — APPROVED** [review]
Thanks!

**Pull request merged**